In [33]:
import pandas as pd

In [34]:
data = pd.read_csv('~/Desktop/2017-22_complete.csv')

In [35]:
data

MATCH UP  GAME DATE  W/L  MIN  eFG%  FTARate  TOV%  OREB%  OppeFG%  \
0        DEN @ GSW  9/30/2017    W   48  52.5     0.42  16.7     22     43.2   
1      LAL vs. MIN  9/30/2017    L   48  46.8    0.183  16.1   28.3     44.9   
2      GSW vs. DEN  9/30/2017    L   48  43.2    0.253    13     31     52.5   
3        MIN @ LAL  9/30/2017    W   48  44.9    0.393  16.1   34.5     46.8   
4        LAC @ TOR  10/1/2017    L   48  50.6    0.571  17.4   13.6     57.1   
...            ...        ...  ...  ...   ...      ...   ...    ...      ...   
10776    BOS @ GSW  6/13/2022    L   48  48.7    0.413  18.4     25     51.7   
10777  GSW vs. BOS  6/13/2022    W   48  51.7     0.17   7.2   10.6     48.7   
10778    GSW @ BOS  6/16/2022    W   48  51.6    0.087  18.1   40.7     49.4   
10779  BOS vs. GSW  6/16/2022    L   48  49.4     0.15  24.7   34.8     51.6   
10780     MATCH UP  GAME DATE  W/L  MIN  eFG%  FTARate  TOV%  OREB%  OppeFG%   

            OppFTA  OppTOV%  OppOREB%  
0            0.253       13        31  
1            0.393     16.1      34.5  
2             0.42     16.7        22  
3            0.183     16.1      28.3  
4            0.513     18.8      23.3  
...            ...      ...       ...  
10776         0.17      7.2      10.6  
10777        0.413     18.4        25  
10778         0.15     24.7      34.8  
10779        0.087     18.1      40.7  
10780  OppFTA Rate  OppTOV%  OppOREB%  

[10781 rows x 12 columns]

In [36]:
data = data[data['MATCH UP'] != 'MATCH UP']

In [37]:
data

MATCH UP  GAME DATE W/L MIN  eFG% FTARate  TOV% OREB% OppeFG%  \
0        DEN @ GSW  9/30/2017   W  48  52.5    0.42  16.7    22    43.2   
1      LAL vs. MIN  9/30/2017   L  48  46.8   0.183  16.1  28.3    44.9   
2      GSW vs. DEN  9/30/2017   L  48  43.2   0.253    13    31    52.5   
3        MIN @ LAL  9/30/2017   W  48  44.9   0.393  16.1  34.5    46.8   
4        LAC @ TOR  10/1/2017   L  48  50.6   0.571  17.4  13.6    57.1   
...            ...        ...  ..  ..   ...     ...   ...   ...     ...   
10775    GSW @ BOS  6/10/2022   W  48  52.2   0.165  16.7  38.5    48.8   
10776    BOS @ GSW  6/13/2022   L  48  48.7   0.413  18.4    25    51.7   
10777  GSW vs. BOS  6/13/2022   W  48  51.7    0.17   7.2  10.6    48.7   
10778    GSW @ BOS  6/16/2022   W  48  51.6   0.087  18.1  40.7    49.4   
10779  BOS vs. GSW  6/16/2022   L  48  49.4    0.15  24.7  34.8    51.6   

      OppFTA OppTOV% OppOREB%  
0      0.253      13       31  
1      0.393    16.1     34.5  
2       0.42    16.7       22  
3      0.183    16.1     28.3  
4      0.513    18.8     23.3  
...      ...     ...      ...  
10775  0.224    16.7     26.4  
10776   0.17     7.2     10.6  
10777  0.413    18.4       25  
10778   0.15    24.7     34.8  
10779  0.087    18.1     40.7  

[10780 rows x 12 columns]

In [38]:
def extract_first_team(matchup):
    return matchup.split()[0]

# Apply the function to create a new column 'Team'
data['Team'] = data['MATCH UP'].apply(extract_first_team)

/var/folders/q9/m0krt3q93nd2zg9qk8_03kk40000gn/T/ipykernel_96493/1401709533.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Team'] = data['MATCH UP'].apply(extract_first_team)


In [39]:
data

MATCH UP  GAME DATE W/L MIN  eFG% FTARate  TOV% OREB% OppeFG%  \
0        DEN @ GSW  9/30/2017   W  48  52.5    0.42  16.7    22    43.2   
1      LAL vs. MIN  9/30/2017   L  48  46.8   0.183  16.1  28.3    44.9   
2      GSW vs. DEN  9/30/2017   L  48  43.2   0.253    13    31    52.5   
3        MIN @ LAL  9/30/2017   W  48  44.9   0.393  16.1  34.5    46.8   
4        LAC @ TOR  10/1/2017   L  48  50.6   0.571  17.4  13.6    57.1   
...            ...        ...  ..  ..   ...     ...   ...   ...     ...   
10775    GSW @ BOS  6/10/2022   W  48  52.2   0.165  16.7  38.5    48.8   
10776    BOS @ GSW  6/13/2022   L  48  48.7   0.413  18.4    25    51.7   
10777  GSW vs. BOS  6/13/2022   W  48  51.7    0.17   7.2  10.6    48.7   
10778    GSW @ BOS  6/16/2022   W  48  51.6   0.087  18.1  40.7    49.4   
10779  BOS vs. GSW  6/16/2022   L  48  49.4    0.15  24.7  34.8    51.6   

      OppFTA OppTOV% OppOREB% Team  
0      0.253      13       31  DEN  
1      0.393    16.1     34.5  LAL  
2       0.42    16.7       22  GSW  
3      0.183    16.1     28.3  MIN  
4      0.513    18.8     23.3  LAC  
...      ...     ...      ...  ...  
10775  0.224    16.7     26.4  GSW  
10776   0.17     7.2     10.6  BOS  
10777  0.413    18.4       25  GSW  
10778   0.15    24.7     34.8  GSW  
10779  0.087    18.1     40.7  BOS  

[10780 rows x 13 columns]

In [40]:
mean_values = data.groupby('Team').agg({
    'eFG%': 'mean',
    'FTARate': 'mean',
    'TOV%': 'mean',
    'OREB%': 'mean',
    'OppFTA': 'mean',
    'OppeFG%': 'mean',
    'OppTOV%': 'mean',
    'OppOREB%': 'mean'
}).reset_index()

mean_values

TypeError: Could not convert 39.155.644.751.944.455.941.637.85246.349.455.247.954.954.156.253.949.455.773.453.248.254.25438.759.552.542.148.961.953.955.453.651.950.648.160.662.749.548.352.847.1415044.857.553.151.747.24455.443.156.35057.658.449.448.955.449.454.255.145.346.149.441.748.952.954.846.347.662.85055.652.945.246.450.650.661.740.154.749.441.755.158.748.549.553.645.666.752.251.157.860.248.939.341.562.852.959.751.846.2505051.739.740.250.550.546.961.75540.350.551.751.151.152.251.263.554.452.742.960.155495047.451.852.751.739.46346.972.25050.658.352.256.353.546.959.561.55561.654.452.241.860.247.859.448.158.55951.146.24157.758.750.546.255.45853.355.245.253.452.158.748.948.55042.452.244.848.862.5506156.654.847.651.741.146.1464447.153.848.554.85952.554.851.754.557.837.269.653.449.452.848.862.347.859.352.866.550.5465757.151.758.353.56160.346.451.652.462.555.355.956.563.850.661.449.560.155.455.453.55646.353.963.152.841.2375059.564.370.85953.15751.158.353.34461.75347.762.852.451.242.651.246.748.853.74755.458.253.947.845.345.645.957.955.943.457.845.85842.153.650.5475056.754.143.556.263.558.659.856.35865.740.360.547.460.858.345.258.255.950.647.549.445.453.453.852.662.55051.649.445.758.26052.96166.2546052.864.353.164.351.141.667.252.44362.56046.465.747.663.148.951.156.452.864.942.25056.652.958.647.859.664.563.758.849.453.255.452.363.860.654.445.35458.85047.3 to numeric

In [14]:
data.columns

Index(['MATCH UP', 'GAME DATE', 'W/L', 'MIN', 'eFG%', 'FTARate', 'TOV%',
       'OREB%', 'OppeFG%', 'OppFTA Rate', 'OppTOV%', 'OppOREB%', 'Team'],
      dtype='object')

In [17]:
data_types = data.dtypes

data_types

MATCH UP       object
GAME DATE      object
W/L            object
MIN            object
eFG%           object
FTARate        object
TOV%           object
OREB%          object
OppeFG%        object
OppFTA Rate    object
OppTOV%        object
OppOREB%       object
Team           object
dtype: object

In [41]:
columns_to_convert = ['eFG%', 'FTARate', 'TOV%', 'OREB%', 'OppeFG%', 'OppFTA', 'OppTOV%', 'OppOREB%']

for column in columns_to_convert:
    data[column] = pd.to_numeric(data[column], errors='coerce')

/var/folders/q9/m0krt3q93nd2zg9qk8_03kk40000gn/T/ipykernel_96493/1167119023.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[column] = pd.to_numeric(data[column], errors='coerce')


In [46]:
def getResult(winloss):
    if winloss == 'W':
        return 1
    return 0

data['win'] = data['W/L'].apply(getResult)

data

/var/folders/q9/m0krt3q93nd2zg9qk8_03kk40000gn/T/ipykernel_96493/3757682544.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['win'] = data['W/L'].apply(getResult)


MATCH UP  GAME DATE W/L MIN  eFG%  FTARate  TOV%  OREB%  OppeFG%  \
0        DEN @ GSW  9/30/2017   W  48  52.5    0.420  16.7   22.0     43.2   
1      LAL vs. MIN  9/30/2017   L  48  46.8    0.183  16.1   28.3     44.9   
2      GSW vs. DEN  9/30/2017   L  48  43.2    0.253  13.0   31.0     52.5   
3        MIN @ LAL  9/30/2017   W  48  44.9    0.393  16.1   34.5     46.8   
4        LAC @ TOR  10/1/2017   L  48  50.6    0.571  17.4   13.6     57.1   
...            ...        ...  ..  ..   ...      ...   ...    ...      ...   
10775    GSW @ BOS  6/10/2022   W  48  52.2    0.165  16.7   38.5     48.8   
10776    BOS @ GSW  6/13/2022   L  48  48.7    0.413  18.4   25.0     51.7   
10777  GSW vs. BOS  6/13/2022   W  48  51.7    0.170   7.2   10.6     48.7   
10778    GSW @ BOS  6/16/2022   W  48  51.6    0.087  18.1   40.7     49.4   
10779  BOS vs. GSW  6/16/2022   L  48  49.4    0.150  24.7   34.8     51.6   

       OppFTA  OppTOV%  OppOREB% Team  win  
0       0.253     13.0      31.0  DEN    1  
1       0.393     16.1      34.5  LAL    0  
2       0.420     16.7      22.0  GSW    0  
3       0.183     16.1      28.3  MIN    1  
4       0.513     18.8      23.3  LAC    0  
...       ...      ...       ...  ...  ...  
10775   0.224     16.7      26.4  GSW    1  
10776   0.170      7.2      10.6  BOS    0  
10777   0.413     18.4      25.0  GSW    1  
10778   0.150     24.7      34.8  GSW    1  
10779   0.087     18.1      40.7  BOS    0  

[10780 rows x 14 columns]

In [59]:
mean_values = data.groupby('Team').agg({
    'eFG%': 'mean',
    'FTARate': 'mean',
    'TOV%': 'mean',
    'OREB%': 'mean',
    'OppeFG%': 'mean',
    'OppFTA': 'mean',
    'OppTOV%': 'mean',
    'OppOREB%': 'mean',
    'win': 'sum'
}).reset_index()

mean_values

Team       eFG%   FTARate       TOV%      OREB%    OppeFG%    OppFTA  \
0   ADL  44.000000  0.476000  17.200000  15.300000  57.700000  0.247000   
1   ATL  52.756787  0.258861  14.445429  26.980886  53.642105  0.258283   
2   BJD  28.200000  0.437000  18.500000  23.200000  62.000000  0.480000   
3   BKN  53.792373  0.267966  14.416949  26.344633  51.977684  0.257062   
4   BNE  42.400000  0.162000  13.800000  31.000000  52.300000  0.295000   
5   BOS  52.972308  0.243872  13.827436  26.800513  51.196154  0.259995   
6   CHA  52.438757  0.265080  13.371598  26.736391  54.155325  0.228183   
7   CHI  52.068421  0.225871  14.026316  24.906725  53.945614  0.249898   
8   CLE  52.368889  0.264164  14.691667  26.859444  54.323333  0.232925   
9   DAL  53.164444  0.256511  13.007500  25.052222  52.784444  0.254061   
10  DEN  54.167123  0.245518  14.181370  28.945753  53.536712  0.254879   
11  DET  50.761471  0.254115  14.250588  27.227647  53.354412  0.265932   
12  FLA  36.300000  0.286000  16.000000  20.000000  60.500000  0.272000   
13  GSW  55.759701  0.240721  14.712687  25.264677  51.028856  0.270806   
14  GUA  45.800000  0.401500  14.450000  30.100000  64.150000  0.265000   
15  HOU  53.769231  0.283126  14.607418  25.623901  53.548077  0.261135   
16  IND  53.072543  0.235471  13.926301  27.193931  53.179191  0.252263   
17  LAC  53.717680  0.276961  14.076796  26.448895  52.181768  0.257077   
18  LAL  52.717579  0.270058  15.135447  26.366571  52.391354  0.253591   
19  MAC  41.940000  0.217600  20.760000  24.400000  59.640000  0.352600   
20  MEL  40.500000  0.251333  19.166667  29.533333  49.000000  0.364000   
21  MEM  51.433894  0.257868  14.161064  28.250700  52.577871  0.288050   
22  MIA  52.947397  0.252584  14.849863  26.584658  52.099178  0.266934   
23  MIL  54.448849  0.259266  13.579795  25.745013  52.296419  0.234790   
24  MIN  52.276522  0.272339  13.586667  27.968116  54.187826  0.265417   
25  NOP  53.076571  0.266183  14.410000  27.946286  53.576857  0.242586   
26  NYK  50.835484  0.257032  14.122874  27.465396  52.200880  0.267493   
27  NZB  39.100000  0.322000  14.300000  27.400000  43.500000  0.403000   
28  OKC  50.829070  0.256273  14.407267  27.949709  53.061337  0.245445   
29  ORL  50.543275  0.231901  13.969883  24.920468  52.784795  0.241558   
30  PER  36.550000  0.350500  18.300000  26.500000  58.300000  0.287500   
31  PHI  53.590501  0.296449  14.645383  27.138786  51.227177  0.276084   
32  PHX  53.167655  0.247720  14.195418  25.377358  52.897844  0.279313   
33  POR  52.345179  0.252251  13.551515  27.946832  53.019008  0.269047   
34  SAC  52.343787  0.240731  13.692604  25.897337  54.547337  0.257204   
35  SAS  52.078000  0.240646  12.521429  25.988000  52.555714  0.230017   
36  SDS  49.050000  0.166500  25.600000  14.750000  70.350000  0.178500   
37  SYD  46.000000  0.238000  22.500000  28.100000  57.050000  0.278000   
38  TOR  53.069415  0.250886  13.434309  27.363830  52.148404  0.275423   
39  UTA  54.535326  0.273019  14.706250  27.500543  51.200815  0.239171   
40  WAS  52.965230  0.268509  13.954598  25.459770  53.440805  0.272388   

      OppTOV%   OppOREB%  win  
0   17.100000  25.000000    0  
1   13.867590  27.139889  158  
2   14.100000  40.000000    0  
3   13.027966  27.980226  181  
4   11.900000  18.200000    0  
5   14.376667  26.556154  230  
6   14.239349  26.609763  159  
7   13.360526  24.622222  139  
8   13.668056  26.704722  154  
9   13.607222  26.047778  175  
10  13.623288  25.340548  217  
11  14.650294  26.699118  131  
12  13.000000  15.800000    0  
13  14.323632  27.188557  263  
14  15.200000  39.650000    0  
15  14.743407  27.481593  184  
16  15.027746  27.690173  166  
17  13.519613  27.435912  199  
18  14.291931  26.646686  159  
19  16.240000  29.920000    0  
20  15.733333  30.833333    0  
21  15.315406  26.973109  169  
22  15.075342  26.305753  200  
23  13.546036  25.307673  242  
24  15.199420  28.319710  163  
25  13.606571  25.80

In [60]:
mean_values = mean_values[mean_values['win'] > 0]

In [61]:
mean_values

Team       eFG%   FTARate       TOV%      OREB%    OppeFG%    OppFTA  \
1   ATL  52.756787  0.258861  14.445429  26.980886  53.642105  0.258283   
3   BKN  53.792373  0.267966  14.416949  26.344633  51.977684  0.257062   
5   BOS  52.972308  0.243872  13.827436  26.800513  51.196154  0.259995   
6   CHA  52.438757  0.265080  13.371598  26.736391  54.155325  0.228183   
7   CHI  52.068421  0.225871  14.026316  24.906725  53.945614  0.249898   
8   CLE  52.368889  0.264164  14.691667  26.859444  54.323333  0.232925   
9   DAL  53.164444  0.256511  13.007500  25.052222  52.784444  0.254061   
10  DEN  54.167123  0.245518  14.181370  28.945753  53.536712  0.254879   
11  DET  50.761471  0.254115  14.250588  27.227647  53.354412  0.265932   
13  GSW  55.759701  0.240721  14.712687  25.264677  51.028856  0.270806   
15  HOU  53.769231  0.283126  14.607418  25.623901  53.548077  0.261135   
16  IND  53.072543  0.235471  13.926301  27.193931  53.179191  0.252263   
17  LAC  53.717680  0.276961  14.076796  26.448895  52.181768  0.257077   
18  LAL  52.717579  0.270058  15.135447  26.366571  52.391354  0.253591   
21  MEM  51.433894  0.257868  14.161064  28.250700  52.577871  0.288050   
22  MIA  52.947397  0.252584  14.849863  26.584658  52.099178  0.266934   
23  MIL  54.448849  0.259266  13.579795  25.745013  52.296419  0.234790   
24  MIN  52.276522  0.272339  13.586667  27.968116  54.187826  0.265417   
25  NOP  53.076571  0.266183  14.410000  27.946286  53.576857  0.242586   
26  NYK  50.835484  0.257032  14.122874  27.465396  52.200880  0.267493   
28  OKC  50.829070  0.256273  14.407267  27.949709  53.061337  0.245445   
29  ORL  50.543275  0.231901  13.969883  24.920468  52.784795  0.241558   
31  PHI  53.590501  0.296449  14.645383  27.138786  51.227177  0.276084   
32  PHX  53.167655  0.247720  14.195418  25.377358  52.897844  0.279313   
33  POR  52.345179  0.252251  13.551515  27.946832  53.019008  0.269047   
34  SAC  52.343787  0.240731  13.692604  25.897337  54.547337  0.257204   
35  SAS  52.078000  0.240646  12.521429  25.988000  52.555714  0.230017   
38  TOR  53.069415  0.250886  13.434309  27.363830  52.148404  0.275423   
39  UTA  54.535326  0.273019  14.706250  27.500543  51.200815  0.239171   
40  WAS  52.965230  0.268509  13.954598  25.459770  53.440805  0.272388   

      OppTOV%   OppOREB%  win  
1   13.867590  27.139889  158  
3   13.027966  27.980226  181  
5   14.376667  26.556154  230  
6   14.239349  26.609763  159  
7   13.360526  24.622222  139  
8   13.668056  26.704722  154  
9   13.607222  26.047778  175  
10  13.623288  25.340548  217  
11  14.650294  26.699118  131  
13  14.323632  27.188557  263  
15  14.743407  27.481593  184  
16  15.027746  27.690173  166  
17  13.519613  27.435912  199  
18  14.291931  26.646686  159  
21  15.315406  26.973109  169  
22  15.075342  26.305753  200  
23  13.546036  25.307673  242  
24  15.199420  28.319710  163  
25  13.606571  25.804857  161  
26  13.448387  26.508798  135  
28  14.718895  26.512500  154  
29  13.518421  25.982456  119  
31  13.851715  26.340633  237  
32  14.132345  27.872507  183  
33  13.095317  26.476584  190  
34  14.439645  27.797929  136  
35  13.406857  26.647429  175  
38  15.557447  27.581649  226  
39  13.364674  24.434783  228  
40  14.154598  26.942241  157

In [62]:
len(mean_values)

30

In [63]:
mean_values['win'] = mean_values['win'] / 4

/var/folders/q9/m0krt3q93nd2zg9qk8_03kk40000gn/T/ipykernel_96493/2815672039.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mean_values['win'] = mean_values['win'] / 4


In [64]:
mean_values

Team       eFG%   FTARate       TOV%      OREB%    OppeFG%    OppFTA  \
1   ATL  52.756787  0.258861  14.445429  26.980886  53.642105  0.258283   
3   BKN  53.792373  0.267966  14.416949  26.344633  51.977684  0.257062   
5   BOS  52.972308  0.243872  13.827436  26.800513  51.196154  0.259995   
6   CHA  52.438757  0.265080  13.371598  26.736391  54.155325  0.228183   
7   CHI  52.068421  0.225871  14.026316  24.906725  53.945614  0.249898   
8   CLE  52.368889  0.264164  14.691667  26.859444  54.323333  0.232925   
9   DAL  53.164444  0.256511  13.007500  25.052222  52.784444  0.254061   
10  DEN  54.167123  0.245518  14.181370  28.945753  53.536712  0.254879   
11  DET  50.761471  0.254115  14.250588  27.227647  53.354412  0.265932   
13  GSW  55.759701  0.240721  14.712687  25.264677  51.028856  0.270806   
15  HOU  53.769231  0.283126  14.607418  25.623901  53.548077  0.261135   
16  IND  53.072543  0.235471  13.926301  27.193931  53.179191  0.252263   
17  LAC  53.717680  0.276961  14.076796  26.448895  52.181768  0.257077   
18  LAL  52.717579  0.270058  15.135447  26.366571  52.391354  0.253591   
21  MEM  51.433894  0.257868  14.161064  28.250700  52.577871  0.288050   
22  MIA  52.947397  0.252584  14.849863  26.584658  52.099178  0.266934   
23  MIL  54.448849  0.259266  13.579795  25.745013  52.296419  0.234790   
24  MIN  52.276522  0.272339  13.586667  27.968116  54.187826  0.265417   
25  NOP  53.076571  0.266183  14.410000  27.946286  53.576857  0.242586   
26  NYK  50.835484  0.257032  14.122874  27.465396  52.200880  0.267493   
28  OKC  50.829070  0.256273  14.407267  27.949709  53.061337  0.245445   
29  ORL  50.543275  0.231901  13.969883  24.920468  52.784795  0.241558   
31  PHI  53.590501  0.296449  14.645383  27.138786  51.227177  0.276084   
32  PHX  53.167655  0.247720  14.195418  25.377358  52.897844  0.279313   
33  POR  52.345179  0.252251  13.551515  27.946832  53.019008  0.269047   
34  SAC  52.343787  0.240731  13.692604  25.897337  54.547337  0.257204   
35  SAS  52.078000  0.240646  12.521429  25.988000  52.555714  0.230017   
38  TOR  53.069415  0.250886  13.434309  27.363830  52.148404  0.275423   
39  UTA  54.535326  0.273019  14.706250  27.500543  51.200815  0.239171   
40  WAS  52.965230  0.268509  13.954598  25.459770  53.440805  0.272388   

      OppTOV%   OppOREB%    win  
1   13.867590  27.139889  39.50  
3   13.027966  27.980226  45.25  
5   14.376667  26.556154  57.50  
6   14.239349  26.609763  39.75  
7   13.360526  24.622222  34.75  
8   13.668056  26.704722  38.50  
9   13.607222  26.047778  43.75  
10  13.623288  25.340548  54.25  
11  14.650294  26.699118  32.75  
13  14.323632  27.188557  65.75  
15  14.743407  27.481593  46.00  
16  15.027746  27.690173  41.50  
17  13.519613  27.435912  49.75  
18  14.291931  26.646686  39.75  
21  15.315406  26.973109  42.25  
22  15.075342  26.305753  50.00  
23  13.546036  25.307673  60.50  
24  15.199420  28.319710  40.75  
25  13.606571  25.804857  40.25  
26  13.448387  26.508798  33.75  
28  14.718895  26.512500  38.50  
29  13.518421  25.982456  29.75  
31  13.851715  26.340633  59.25  
32  14.132345  27.872507  45.75  
33  13.095317  26.476584  47.50  
34  14.439645  27.797929  34.00  
35  13.406857  26.647429  43.75  
38  15.557447  27.581649  56.50  
39  13.364674  24.434783  57.00  
40  14.154598  26.942241  39.25

In [65]:
vals = mean_values

In [66]:
vals

Team       eFG%   FTARate       TOV%      OREB%    OppeFG%    OppFTA  \
1   ATL  52.756787  0.258861  14.445429  26.980886  53.642105  0.258283   
3   BKN  53.792373  0.267966  14.416949  26.344633  51.977684  0.257062   
5   BOS  52.972308  0.243872  13.827436  26.800513  51.196154  0.259995   
6   CHA  52.438757  0.265080  13.371598  26.736391  54.155325  0.228183   
7   CHI  52.068421  0.225871  14.026316  24.906725  53.945614  0.249898   
8   CLE  52.368889  0.264164  14.691667  26.859444  54.323333  0.232925   
9   DAL  53.164444  0.256511  13.007500  25.052222  52.784444  0.254061   
10  DEN  54.167123  0.245518  14.181370  28.945753  53.536712  0.254879   
11  DET  50.761471  0.254115  14.250588  27.227647  53.354412  0.265932   
13  GSW  55.759701  0.240721  14.712687  25.264677  51.028856  0.270806   
15  HOU  53.769231  0.283126  14.607418  25.623901  53.548077  0.261135   
16  IND  53.072543  0.235471  13.926301  27.193931  53.179191  0.252263   
17  LAC  53.717680  0.276961  14.076796  26.448895  52.181768  0.257077   
18  LAL  52.717579  0.270058  15.135447  26.366571  52.391354  0.253591   
21  MEM  51.433894  0.257868  14.161064  28.250700  52.577871  0.288050   
22  MIA  52.947397  0.252584  14.849863  26.584658  52.099178  0.266934   
23  MIL  54.448849  0.259266  13.579795  25.745013  52.296419  0.234790   
24  MIN  52.276522  0.272339  13.586667  27.968116  54.187826  0.265417   
25  NOP  53.076571  0.266183  14.410000  27.946286  53.576857  0.242586   
26  NYK  50.835484  0.257032  14.122874  27.465396  52.200880  0.267493   
28  OKC  50.829070  0.256273  14.407267  27.949709  53.061337  0.245445   
29  ORL  50.543275  0.231901  13.969883  24.920468  52.784795  0.241558   
31  PHI  53.590501  0.296449  14.645383  27.138786  51.227177  0.276084   
32  PHX  53.167655  0.247720  14.195418  25.377358  52.897844  0.279313   
33  POR  52.345179  0.252251  13.551515  27.946832  53.019008  0.269047   
34  SAC  52.343787  0.240731  13.692604  25.897337  54.547337  0.257204   
35  SAS  52.078000  0.240646  12.521429  25.988000  52.555714  0.230017   
38  TOR  53.069415  0.250886  13.434309  27.363830  52.148404  0.275423   
39  UTA  54.535326  0.273019  14.706250  27.500543  51.200815  0.239171   
40  WAS  52.965230  0.268509  13.954598  25.459770  53.440805  0.272388   

      OppTOV%   OppOREB%    win  
1   13.867590  27.139889  39.50  
3   13.027966  27.980226  45.25  
5   14.376667  26.556154  57.50  
6   14.239349  26.609763  39.75  
7   13.360526  24.622222  34.75  
8   13.668056  26.704722  38.50  
9   13.607222  26.047778  43.75  
10  13.623288  25.340548  54.25  
11  14.650294  26.699118  32.75  
13  14.323632  27.188557  65.75  
15  14.743407  27.481593  46.00  
16  15.027746  27.690173  41.50  
17  13.519613  27.435912  49.75  
18  14.291931  26.646686  39.75  
21  15.315406  26.973109  42.25  
22  15.075342  26.305753  50.00  
23  13.546036  25.307673  60.50  
24  15.199420  28.319710  40.75  
25  13.606571  25.804857  40.25  
26  13.448387  26.508798  33.75  
28  14.718895  26.512500  38.50  
29  13.518421  25.982456  29.75  
31  13.851715  26.340633  59.25  
32  14.132345  27.872507  45.75  
33  13.095317  26.476584  47.50  
34  14.439645  27.797929  34.00  
35  13.406857  26.647429  43.75  
38  15.557447  27.581649  56.50  
39  13.364674  24.434783  57.00  
40  14.154598  26.942241  39.25

In [72]:
vals['fgdiff'] = vals['eFG%'] - vals['OppeFG%']
vals['tovdiff'] = vals['TOV%'] - vals['OppTOV%']
vals['ftdiff'] = vals['FTARate'] - vals['OppFTA']
vals['rebdiff'] = vals['OREB%'] - vals['OppOREB%']

/var/folders/q9/m0krt3q93nd2zg9qk8_03kk40000gn/T/ipykernel_96493/3012747336.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  vals['fgdiff'] = vals['eFG%'] - vals['OppeFG%']
/var/folders/q9/m0krt3q93nd2zg9qk8_03kk40000gn/T/ipykernel_96493/3012747336.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  vals['tovdiff'] = vals['TOV%'] - vals['OppTOV%']
/var/folders/q9/m0krt3q93nd2zg9qk8_03kk40000gn/T/ipykernel_96493/3012747336.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice fro

In [73]:
vals

Team       eFG%   FTARate       TOV%      OREB%    OppeFG%    OppFTA  \
1   ATL  52.756787  0.258861  14.445429  26.980886  53.642105  0.258283   
3   BKN  53.792373  0.267966  14.416949  26.344633  51.977684  0.257062   
5   BOS  52.972308  0.243872  13.827436  26.800513  51.196154  0.259995   
6   CHA  52.438757  0.265080  13.371598  26.736391  54.155325  0.228183   
7   CHI  52.068421  0.225871  14.026316  24.906725  53.945614  0.249898   
8   CLE  52.368889  0.264164  14.691667  26.859444  54.323333  0.232925   
9   DAL  53.164444  0.256511  13.007500  25.052222  52.784444  0.254061   
10  DEN  54.167123  0.245518  14.181370  28.945753  53.536712  0.254879   
11  DET  50.761471  0.254115  14.250588  27.227647  53.354412  0.265932   
13  GSW  55.759701  0.240721  14.712687  25.264677  51.028856  0.270806   
15  HOU  53.769231  0.283126  14.607418  25.623901  53.548077  0.261135   
16  IND  53.072543  0.235471  13.926301  27.193931  53.179191  0.252263   
17  LAC  53.717680  0.276961  14.076796  26.448895  52.181768  0.257077   
18  LAL  52.717579  0.270058  15.135447  26.366571  52.391354  0.253591   
21  MEM  51.433894  0.257868  14.161064  28.250700  52.577871  0.288050   
22  MIA  52.947397  0.252584  14.849863  26.584658  52.099178  0.266934   
23  MIL  54.448849  0.259266  13.579795  25.745013  52.296419  0.234790   
24  MIN  52.276522  0.272339  13.586667  27.968116  54.187826  0.265417   
25  NOP  53.076571  0.266183  14.410000  27.946286  53.576857  0.242586   
26  NYK  50.835484  0.257032  14.122874  27.465396  52.200880  0.267493   
28  OKC  50.829070  0.256273  14.407267  27.949709  53.061337  0.245445   
29  ORL  50.543275  0.231901  13.969883  24.920468  52.784795  0.241558   
31  PHI  53.590501  0.296449  14.645383  27.138786  51.227177  0.276084   
32  PHX  53.167655  0.247720  14.195418  25.377358  52.897844  0.279313   
33  POR  52.345179  0.252251  13.551515  27.946832  53.019008  0.269047   
34  SAC  52.343787  0.240731  13.692604  25.897337  54.547337  0.257204   
35  SAS  52.078000  0.240646  12.521429  25.988000  52.555714  0.230017   
38  TOR  53.069415  0.250886  13.434309  27.363830  52.148404  0.275423   
39  UTA  54.535326  0.273019  14.706250  27.500543  51.200815  0.239171   
40  WAS  52.965230  0.268509  13.954598  25.459770  53.440805  0.272388   

      OppTOV%   OppOREB%    win    fgdiff   tovdiff    ftdiff   rebdiff  
1   13.867590  27.139889  39.50 -0.885319  0.577839  0.000579 -0.159003  
3   13.027966  27.980226  45.25  1.814689  1.388983  0.010904 -1.635593  
5   14.376667  26.556154  57.50  1.776154 -0.549231 -0.016123  0.244359  
6   14.239349  26.609763  39.75 -1.716568 -0.867751  0.036896  0.126627  
7   13.360526  24.622222  34.75 -1.877193  0.665789 -0.024026  0.284503  
8   13.668056  26.704722  38.50 -1.954444  1.023611  0.031239  0.154722  
9   13.607222  26.047778  43.75  0.380000 -0.599722  0.002450 -0.995556  
10  13.623288  25.340548  54.25  0.630411  0.558082 -0.009362  3.605205  
11  14.650294  26.699118  32.75 -2.592941 -0.399706 -0.011818  0.528529  
13  14.323632  27.188557  65.75  4.730846  0.389055 -0.030085 -1.923881  
15  14.743407  27.481593  46.00  0.221154 -0.135989  0.021992 -1.857692  
16  15.027746  27.690173  41.50 -0.106647 -1.101445 -0.016792 -0.496243  
17  13.519613  27.435912  49.75  1.535912  0.557182  0.019884 -0.987017  
18  14.291931  26.646686  39.75  0.326225  0.843516  0.016467 -0.280115  
21  15.315406  26.973109  42.25 -1.143978 -1.154342 -0.030182  1.277591  
22  15.075342  26.305753  50.00  0.848219 -0.225479 -0.014351  0.278904  
23  13.546036  25.307673  60.50  2.152430  0.033760  0.024476  0.437340  
24  15.199420  28.319710  40.75 -1.911304 -1.612754  0.006922 -0.351594  
25  13.606571  25.804857  40.25 -0.500286  0.803429  0.023597  2.141429  
26  13.448387  26.508798  33.75 -1.365396  0.674487 -0.010460  0.956598  
28  14.718895  26.512500  38.50 -2.232267 -0.311628  0.010828  1.437209  
29  13.518421  25.982456  29.75 -2.241520  0.4514

In [74]:
vals.columns

Index(['Team', 'eFG%', 'FTARate', 'TOV%', 'OREB%', 'OppeFG%', 'OppFTA',
       'OppTOV%', 'OppOREB%', 'win', 'fgdiff', 'tovdiff', 'ftdiff', 'rebdiff'],
      dtype='object')

In [75]:
x = vals[['fgdiff', 'tovdiff', 'ftdiff', 'rebdiff']]
y = vals['win']

In [76]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

# Create and train the model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

/Users/anirudhkaranam/opt/anaconda3/lib/python3.9/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.26.4
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [77]:
print("Mean Squared Error:", mean_squared_error(y_test, y_pred))
print("R-squared:", r2_score(y_test, y_pred))

# Display model coefficients
print("Coefficients:", model.coef_)
print("Intercept:", model.intercept_)

Mean Squared Error: 8.840488706125146
R-squared: 0.9219335025644189
Coefficients: [ 4.63877617 -2.72506519 26.85466581  1.61272316]
Intercept: 44.62744119944121


### equation is 44.62744119944121 + 4.63877617 * fgdiff - 2.72506519 * tovdiff + 26.85466581 * ftdiff + 1.61272316 * rebdiff

In [96]:
# Let's analyze the four factor model for the 2023-24 season

In [81]:
import pandas as pd

# Create a DataFrame from the provided data
data = {
    'Team': [
        'Boston Celtics', 'Denver Nuggets', 'Oklahoma City Thunder', 'Minnesota Timberwolves', 'LA Clippers',
        'Dallas Mavericks', 'New York Knicks', 'Milwaukee Bucks', 'New Orleans Pelicans', 'Phoenix Suns',
        'Cleveland Cavaliers', 'Indiana Pacers', 'Los Angeles Lakers', 'Orlando Magic', 'Philadelphia 76ers',
        'Golden State Warriors', 'Miami Heat', 'Sacramento Kings', 'Houston Rockets', 'Chicago Bulls',
        'Atlanta Hawks', 'Brooklyn Nets', 'Utah Jazz', 'Memphis Grizzlies', 'Toronto Raptors',
        'San Antonio Spurs', 'Charlotte Hornets', 'Portland Trail Blazers', 'Washington Wizards', 'Detroit Pistons'
    ],
    'GP': [82]*30,
    'W': [64, 57, 57, 56, 51, 50, 50, 49, 49, 49, 48, 47, 47, 47, 47, 46, 46, 46, 41, 39, 36, 32, 31, 27, 25, 22, 21, 21, 15, 14],
    'L': [18, 25, 25, 26, 31, 32, 32, 33, 33, 33, 34, 35, 35, 35, 35, 36, 36, 36, 41, 43, 46, 50, 51, 55, 57, 60, 61, 61, 67, 68],
    'W%': [.780, .695, .695, .683, .622, .610, .610, .598, .598, .598, .585, .573, .573, .573, .573, .561, .561, .561, .500, .476, .439, .390, .378, .329, .305, .268, .256, .256, .183, .171],
    'PTS': [3966, 3941, 3961, 3961, 3941, 3941, 3946, 3961, 3941, 3956, 3961, 3941, 3971, 3956, 3956, 3966, 3951, 3971, 3971, 3996, 3971, 3961, 3961, 3956, 3961, 3966, 3946, 3976, 3946, 3951],
    'eFG%': [57.8, 56.2, 57.3, 55.9, 56.1, 56.2, 54.0, 56.8, 55.8, 56.5, 55.7, 57.8, 56.6, 54.1, 53.2, 55.7, 53.8, 55.6, 52.9, 53.4, 53.9, 53.1, 53.9, 50.9, 53.5, 53.2, 52.9, 50.3, 53.8, 52.6],
    'FTARate': [0.224, 0.224, 0.240, 0.270, 0.256, 0.251, 0.246, 0.270, 0.261, 0.271, 0.234, 0.222, 0.276, 0.287, 0.264, 0.219, 0.257, 0.229, 0.257, 0.236, 0.251, 0.234, 0.251, 0.238, 0.241, 0.220, 0.212, 0.228, 0.221, 0.246],
    'TOV%': [12.1, 13.0, 12.5, 14.4, 13.4, 12.4, 13.8, 12.8, 13.2, 15.0, 13.8, 12.6, 13.7, 15.0, 12.2, 14.2, 13.0, 13.1, 12.7, 12.4, 13.3, 13.4, 15.5, 15.2, 13.9, 14.8, 14.0, 15.4, 13.6, 15.1],
    'OREB%': [28.7, 30.0, 25.0, 27.5, 28.3, 25.7, 33.3, 25.5, 27.8, 29.2, 27.4, 28.4, 24.4, 29.7, 29.2, 31.5, 26.1, 28.5, 29.4, 29.1, 30.8, 28.9, 32.5, 28.7, 28.1, 26.9, 25.3, 32.0, 24.0, 27.7],
    'OppeFG%': [52.3, 52.6, 53.0, 51.5, 54.1, 54.8, 54.3, 53.9, 54.1, 53.6, 53.5, 55.6, 55.0, 54.3, 53.7, 53.9, 54.3, 55.5, 53.3, 55.6, 57.2, 54.3, 56.7, 55.1, 56.6, 55.5, 57.2, 55.8, 56.2, 55.8],
    'OppFTA': [0.189, 0.252, 0.256, 0.252, 0.236, 0.239, 0.231, 0.234, 0.237, 0.230, 0.240, 0.289, 0.192, 0.264, 0.271, 0.248, 0.223, 0.266, 0.286, 0.251, 0.242, 0.239, 0.235, 0.250, 0.218, 0.222, 0.236, 0.273, 0.258, 0.276],
    'OppTOV%': [12.2, 12.7, 15.5, 14.5, 13.3, 13.7, 13.7, 11.9, 14.4, 12.8, 13.8, 13.6, 13.1, 15.4, 14.8, 12.9, 14.1, 13.8, 13.8, 14.2, 13.9, 12.9, 12.2, 15.2, 13.8, 13.1, 13.9, 14.5, 13.6, 12.3],
    'OppOREB%': [27.4, 28.2, 30.2, 27.5, 29.5, 28.9, 27.1, 26.9, 27.5, 28.6, 27.9, 30.0, 27.5, 26.3, 29.9, 27.5, 26.6, 25.6, 27.8, 28.3, 28.3, 28.5, 28.3, 29.4, 30.6, 27.9, 29.1, 30.3, 31.7, 27.0]
}

df = pd.DataFrame(data)

# Drop columns not in the specified list
columns_to_keep = ['Team', 'eFG%', 'FTARate', 'TOV%', 'OREB%', 'OppeFG%', 'OppFTA', 'OppTOV%', 'OppOREB%', 'W']
df = df[columns_to_keep]

# Rename columns
df.columns = ['Team', 'eFG%', 'FTARate', 'TOV%', 'OREB%', 'OppeFG%', 'OppFTA', 'OppTOV%', 'OppOREB%', 'win']

# Display the DataFrame
df


Team  eFG%  FTARate  TOV%  OREB%  OppeFG%  OppFTA  \
0           Boston Celtics  57.8    0.224  12.1   28.7     52.3   0.189   
1           Denver Nuggets  56.2    0.224  13.0   30.0     52.6   0.252   
2    Oklahoma City Thunder  57.3    0.240  12.5   25.0     53.0   0.256   
3   Minnesota Timberwolves  55.9    0.270  14.4   27.5     51.5   0.252   
4              LA Clippers  56.1    0.256  13.4   28.3     54.1   0.236   
5         Dallas Mavericks  56.2    0.251  12.4   25.7     54.8   0.239   
6          New York Knicks  54.0    0.246  13.8   33.3     54.3   0.231   
7          Milwaukee Bucks  56.8    0.270  12.8   25.5     53.9   0.234   
8     New Orleans Pelicans  55.8    0.261  13.2   27.8     54.1   0.237   
9             Phoenix Suns  56.5    0.271  15.0   29.2     53.6   0.230   
10     Cleveland Cavaliers  55.7    0.234  13.8   27.4     53.5   0.240   
11          Indiana Pacers  57.8    0.222  12.6   28.4     55.6   0.289   
12      Los Angeles Lakers  56.6    0.276  13.7   24.4     55.0   0.192   
13           Orlando Magic  54.1    0.287  15.0   29.7     54.3   0.264   
14      Philadelphia 76ers  53.2    0.264  12.2   29.2     53.7   0.271   
15   Golden State Warriors  55.7    0.219  14.2   31.5     53.9   0.248   
16              Miami Heat  53.8    0.257  13.0   26.1     54.3   0.223   
17        Sacramento Kings  55.6    0.229  13.1   28.5     55.5   0.266   
18         Houston Rockets  52.9    0.257  12.7   29.4     53.3   0.286   
19           Chicago Bulls  53.4    0.236  12.4   29.1     55.6   0.251   
20           Atlanta Hawks  53.9    0.251  13.3   30.8     57.2   0.242   
21           Brooklyn Nets  53.1    0.234  13.4   28.9     54.3   0.239   
22               Utah Jazz  53.9    0.251  15.5   32.5     56.7   0.235   
23       Memphis Grizzlies  50.9    0.238  15.2   28.7     55.1   0.250   
24         Toronto Raptors  53.5    0.241  13.9   28.1     56.6   0.218   
25       San Antonio Spurs  53.2    0.220  14.8   26.9     55.5   0.222   
26       Charlotte Hornets  52.9    0.212  14.0   25.3     57.2   0.236   
27  Portland Trail Blazers  50.3    0.228  15.4   32.0     55.8   0.273   
28      Washington Wizards  53.8    0.221  13.6   24.0     56.2   0.258   
29         Detroit Pistons  52.6    0.246  15.1   27.7     55.8   0.276   

    OppTOV%  OppOREB%  win  
0      12.2      27.4   64  
1      12.7      28.2   57  
2      15.5      30.2   57  
3      14.5      27.5   56  
4      13.3      29.5   51  
5      13.7      28.9   50  
6      13.7      27.1   50  
7      11.9      26.9   49  
8      14.4      27.5   49  
9      12.8      28.6   49  
10     13.8      27.9   48  
11     13.6      30.0   47  
12     13.1      27.5   47  
13     15.4      26.3   47  
14     14.8      29.9   47  
15     12.9      27.5   46  
16     14.1      26.6   46  
17     13.8      25.6   46  
18     13.8      27.8   41  
19     14.2      28.3   39  
20     13.9      28.3   36  
21     12.9      28.5   32  
22     12.2      28.3   31  
23     15.2      29.4   27  
24     13.8      30.6   25  
25     13.1      27.9   22  
26     13.9      29.1   21  
27     14.5      30.3   21  
28     13.6      31.7   15  
29     12.3      27.0   14

In [85]:
df['fgdiff'] = df['eFG%'] - df['OppeFG%']
df['tovdiff'] = df['TOV%'] - df['OppTOV%']
df['ftdiff'] = df['FTARate'] - df['OppFTA']
df['rebdiff'] = df['OREB%'] - df['OppOREB%']

In [86]:
df

Team  eFG%  FTARate  TOV%  OREB%  OppeFG%  OppFTA  \
0           Boston Celtics  57.8    0.224  12.1   28.7     52.3   0.189   
1           Denver Nuggets  56.2    0.224  13.0   30.0     52.6   0.252   
2    Oklahoma City Thunder  57.3    0.240  12.5   25.0     53.0   0.256   
3   Minnesota Timberwolves  55.9    0.270  14.4   27.5     51.5   0.252   
4              LA Clippers  56.1    0.256  13.4   28.3     54.1   0.236   
5         Dallas Mavericks  56.2    0.251  12.4   25.7     54.8   0.239   
6          New York Knicks  54.0    0.246  13.8   33.3     54.3   0.231   
7          Milwaukee Bucks  56.8    0.270  12.8   25.5     53.9   0.234   
8     New Orleans Pelicans  55.8    0.261  13.2   27.8     54.1   0.237   
9             Phoenix Suns  56.5    0.271  15.0   29.2     53.6   0.230   
10     Cleveland Cavaliers  55.7    0.234  13.8   27.4     53.5   0.240   
11          Indiana Pacers  57.8    0.222  12.6   28.4     55.6   0.289   
12      Los Angeles Lakers  56.6    0.276  13.7   24.4     55.0   0.192   
13           Orlando Magic  54.1    0.287  15.0   29.7     54.3   0.264   
14      Philadelphia 76ers  53.2    0.264  12.2   29.2     53.7   0.271   
15   Golden State Warriors  55.7    0.219  14.2   31.5     53.9   0.248   
16              Miami Heat  53.8    0.257  13.0   26.1     54.3   0.223   
17        Sacramento Kings  55.6    0.229  13.1   28.5     55.5   0.266   
18         Houston Rockets  52.9    0.257  12.7   29.4     53.3   0.286   
19           Chicago Bulls  53.4    0.236  12.4   29.1     55.6   0.251   
20           Atlanta Hawks  53.9    0.251  13.3   30.8     57.2   0.242   
21           Brooklyn Nets  53.1    0.234  13.4   28.9     54.3   0.239   
22               Utah Jazz  53.9    0.251  15.5   32.5     56.7   0.235   
23       Memphis Grizzlies  50.9    0.238  15.2   28.7     55.1   0.250   
24         Toronto Raptors  53.5    0.241  13.9   28.1     56.6   0.218   
25       San Antonio Spurs  53.2    0.220  14.8   26.9     55.5   0.222   
26       Charlotte Hornets  52.9    0.212  14.0   25.3     57.2   0.236   
27  Portland Trail Blazers  50.3    0.228  15.4   32.0     55.8   0.273   
28      Washington Wizards  53.8    0.221  13.6   24.0     56.2   0.258   
29         Detroit Pistons  52.6    0.246  15.1   27.7     55.8   0.276   

    OppTOV%  OppOREB%  win  fgdiff  tovdiff  ftdiff  rebdiff  
0      12.2      27.4   64     5.5     -0.1   0.035      1.3  
1      12.7      28.2   57     3.6      0.3  -0.028      1.8  
2      15.5      30.2   57     4.3     -3.0  -0.016     -5.2  
3      14.5      27.5   56     4.4     -0.1   0.018      0.0  
4      13.3      29.5   51     2.0      0.1   0.020     -1.2  
5      13.7      28.9   50     1.4     -1.3   0.012     -3.2  
6      13.7      27.1   50    -0.3      0.1   0.015      6.2  
7      11.9      26.9   49     2.9      0.9   0.036     -1.4  
8      14.4      27.5   49     1.7     -1.2   0.024      0.3  
9      12.8      28.6   49     2.9      2.2   0.041      0.6  
10     13.8      27.9   48     2.2      0.0  -0.006     -0.5  
11     13.6      30.0   47     2.2     -1.0  -0.067     -1.6  
12     13.1      27.5   47     1.6      0.6   0.084     -3.1  
13     15.4      26.3   47    -0.2     -0.4   0.023      3.4  
14     14.8      29.9   47    -0.5     -2.6  -0.007     -0.7  
15     12.9      27.5   46     1.8      1.3  -0.029      4.0  
16     14.1      26.6   46    -0.5     -1.1   0.034     -0.5  
17     13.8      25.6   46     0.1     -0.7  -0.037      2.9  
18     13.8      27.8   41    -0.4     -1.1  -0.029      1.6  
19     14.2      28.3   39    -2.2     -1.8  -0.015      0.8  
20     13.9      28.3   36    -3.3     -0.6   0.009      2.5  
21     12.9      28.5   32    -1.2      0.5  -0.005      0.4  
22     12.2      28.3   31    -2.8      3.3   0.016      4.2  
23     15.2      29.4   27    -4.2      0.0  -0.012     -0.7  
24     13.8      30.6   25    -3.1      0.1   0.023     -2.5  
25     13.1      27.9   22    -2.3      1.7  -0.002     -1

In [87]:
x_test = df[['fgdiff', 'tovdiff', 'ftdiff', 'rebdiff']]
y_test = df['win']

model.score(x_test, y_test)

0.8393871875707996

In [88]:
y_pred = model.predict(x_test)

print("Mean Squared Error:", mean_squared_error(y_test, y_pred))
print("R-squared:", r2_score(y_test, y_pred))

Mean Squared Error: 27.925214321023656
R-squared: 0.8393871875707996


##### Our regression model has an MSE of 27.9 and an R^2 value of 0.83. That means we can be ~95% sure that our predicted value for wins will be within 5.283 * 2 = 10.566 wins of the actual value

In [95]:
# Now let's analyze the 2022-23 season

In [89]:
import pandas as pd

# Create the DataFrame
data = {
    'TEAM': [
        'Milwaukee Bucks', 'Boston Celtics', 'Philadelphia 76ers', 'Denver Nuggets', 'Cleveland Cavaliers',
        'Memphis Grizzlies', 'Sacramento Kings', 'New York Knicks', 'Brooklyn Nets', 'Phoenix Suns',
        'Golden State Warriors', 'LA Clippers', 'Miami Heat', 'Los Angeles Lakers', 'Minnesota Timberwolves',
        'New Orleans Pelicans', 'Atlanta Hawks', 'Toronto Raptors', 'Chicago Bulls', 'Oklahoma City Thunder',
        'Dallas Mavericks', 'Utah Jazz', 'Indiana Pacers', 'Washington Wizards', 'Orlando Magic',
        'Portland Trail Blazers', 'Charlotte Hornets', 'Houston Rockets', 'San Antonio Spurs', 'Detroit Pistons'
    ],
    'GP': [82] * 30,
    'W': [58, 57, 54, 53, 51, 51, 48, 47, 45, 45, 44, 44, 44, 43, 42, 42, 41, 41, 40, 40, 38, 37, 35, 35, 34, 33, 27, 22, 22, 17],
    'L': [24, 25, 28, 29, 31, 31, 34, 35, 37, 37, 38, 38, 38, 39, 40, 40, 41, 41, 42, 42, 44, 45, 47, 47, 48, 49, 55, 60, 60, 65],
    'WIN%': [.707, .695, .659, .646, .622, .622, .585, .573, .549, .549, .537, .537, .537, .524, .512, .512, .500, .500, .488, .488, .463, .451, .427, .427, .415, .402, .329, .268, .268, .207],
    'MIN': [3966, 3996, 3976, 3951, 3976, 3956, 3966, 3991, 3946, 3956, 3966, 3966, 3961, 3976, 3966, 3971, 3971, 3961, 3981, 3971, 3986, 3961, 3951, 3951, 3956, 3946, 3966, 3951, 3971, 3961],
    'EFG%': [55.5, 56.6, 56.3, 57.3, 55.6, 54.0, 57.2, 54.1, 56.2, 53.5, 57.1, 55.1, 53.0, 54.2, 56.0, 54.3, 54.1, 51.7, 55.0, 53.1, 56.5, 54.7, 54.5, 55.0, 53.2, 54.9, 51.6, 51.6, 52.5, 52.0],
    'FTA RATE': [0.248, 0.243, 0.300, 0.259, 0.264, 0.259, 0.284, 0.285, 0.260, 0.241, 0.224, 0.278, 0.270, 0.299, 0.271, 0.279, 0.244, 0.257, 0.251, 0.256, 0.298, 0.265, 0.265, 0.258, 0.290, 0.289, 0.261, 0.285, 0.229, 0.295],
    'TOV%': [14.3, 13.3, 13.9, 14.7, 13.7, 13.3, 13.3, 13.1, 13.9, 13.6, 15.8, 14.3, 13.8, 13.7, 15.0, 14.5, 12.6, 11.8, 13.3, 12.7, 12.4, 15.1, 14.6, 14.2, 15.1, 14.5, 13.9, 16.2, 14.8, 15.0],
    'OREB%': [28.4, 25.5, 25.8, 28.9, 27.3, 30.2, 26.5, 31.8, 23.4, 30.7, 27.7, 27.3, 27.4, 26.9, 25.8, 29.0, 28.9, 31.6, 23.6, 28.9, 22.6, 30.8, 28.5, 27.3, 28.1, 26.4, 27.6, 34.4, 29.9, 29.2],
    'OPP EFG%': [52.0, 52.8, 54.1, 54.3, 53.5, 52.6, 56.3, 53.6, 53.0, 53.2, 54.0, 54.3, 56.1, 53.5, 54.0, 54.3, 55.2, 56.5, 54.4, 54.7, 54.9, 54.1, 55.4, 54.0, 55.0, 56.3, 54.4, 56.4, 57.6, 55.7],
    'OPP FTA RATE': [0.225, 0.234, 0.277, 0.260, 0.268, 0.262, 0.258, 0.269, 0.276, 0.298, 0.278, 0.253, 0.253, 0.222, 0.293, 0.272, 0.257, 0.280, 0.253, 0.288, 0.290, 0.265, 0.287, 0.244, 0.265, 0.276, 0.266, 0.275, 0.260, 0.298],
    'OPP TOV%': [11.6, 12.6, 14.5, 13.6, 16.1, 14.8, 13.9, 12.9, 13.9, 14.5, 13.8, 13.0, 16.1, 12.3, 15.0, 14.9, 13.8, 16.9, 15.0, 16.3, 13.3, 12.3, 14.8, 12.3, 14.6, 13.6, 14.0, 13.2, 13.5, 13.4],
    'OPP OREB%': [25.5, 25.4, 27.9, 27.3, 28.5, 28.9, 26.9, 27.3, 31.1, 28.6, 28.1, 26.9, 26.7, 27.3, 29.7, 26.8, 28.3, 27.9, 26.4, 30.7, 28.3, 28.9, 31.5, 27.6, 27.0, 29.1, 28.3, 27.1, 29.1, 29.8],
    'W': [58, 57, 54, 53, 51, 51, 48, 47, 45, 45, 44, 44, 44, 43, 42, 42, 41, 41, 40, 40, 38, 37, 35, 35, 34, 33, 27, 22, 22, 17]
}

df = pd.DataFrame(data)

# Drop unnecessary columns
df = df.drop(columns=['GP', 'L', 'WIN%', 'MIN'])

# Rename columns
df = df.rename(columns={
    'TEAM': 'Team',
    'EFG%': 'eFG%',
    'FTA RATE': 'FTARate',
    'TOV%': 'TOV%',
    'OREB%': 'OREB%',
    'OPP EFG%': 'Opp eFG%',
    'OPP FTA RATE': 'OppFTA',
    'OPP TOV%': 'OppTOV%',
    'OPP OREB%': 'OppOREB%',
    'W': 'win'
})

df


Team  win  eFG%  FTARate  TOV%  OREB%  Opp eFG%  OppFTA  \
0          Milwaukee Bucks   58  55.5    0.248  14.3   28.4      52.0   0.225   
1           Boston Celtics   57  56.6    0.243  13.3   25.5      52.8   0.234   
2       Philadelphia 76ers   54  56.3    0.300  13.9   25.8      54.1   0.277   
3           Denver Nuggets   53  57.3    0.259  14.7   28.9      54.3   0.260   
4      Cleveland Cavaliers   51  55.6    0.264  13.7   27.3      53.5   0.268   
5        Memphis Grizzlies   51  54.0    0.259  13.3   30.2      52.6   0.262   
6         Sacramento Kings   48  57.2    0.284  13.3   26.5      56.3   0.258   
7          New York Knicks   47  54.1    0.285  13.1   31.8      53.6   0.269   
8            Brooklyn Nets   45  56.2    0.260  13.9   23.4      53.0   0.276   
9             Phoenix Suns   45  53.5    0.241  13.6   30.7      53.2   0.298   
10   Golden State Warriors   44  57.1    0.224  15.8   27.7      54.0   0.278   
11             LA Clippers   44  55.1    0.278  14.3   27.3      54.3   0.253   
12              Miami Heat   44  53.0    0.270  13.8   27.4      56.1   0.253   
13      Los Angeles Lakers   43  54.2    0.299  13.7   26.9      53.5   0.222   
14  Minnesota Timberwolves   42  56.0    0.271  15.0   25.8      54.0   0.293   
15    New Orleans Pelicans   42  54.3    0.279  14.5   29.0      54.3   0.272   
16           Atlanta Hawks   41  54.1    0.244  12.6   28.9      55.2   0.257   
17         Toronto Raptors   41  51.7    0.257  11.8   31.6      56.5   0.280   
18           Chicago Bulls   40  55.0    0.251  13.3   23.6      54.4   0.253   
19   Oklahoma City Thunder   40  53.1    0.256  12.7   28.9      54.7   0.288   
20        Dallas Mavericks   38  56.5    0.298  12.4   22.6      54.9   0.290   
21               Utah Jazz   37  54.7    0.265  15.1   30.8      54.1   0.265   
22          Indiana Pacers   35  54.5    0.265  14.6   28.5      55.4   0.287   
23      Washington Wizards   35  55.0    0.258  14.2   27.3      54.0   0.244   
24           Orlando Magic   34  53.2    0.290  15.1   28.1      55.0   0.265   
25  Portland Trail Blazers   33  54.9    0.289  14.5   26.4      56.3   0.276   
26       Charlotte Hornets   27  51.6    0.261  13.9   27.6      54.4   0.266   
27         Houston Rockets   22  51.6    0.285  16.2   34.4      56.4   0.275   
28       San Antonio Spurs   22  52.5    0.229  14.8   29.9      57.6   0.260   
29         Detroit Pistons   17  52.0    0.295  15.0   29.2      55.7   0.298   

    OppTOV%  OppOREB%  
0      11.6      25.5  
1      12.6      25.4  
2      14.5      27.9  
3      13.6      27.3  
4      16.1      28.5  
5      14.8      28.9  
6      13.9      26.9  
7      12.9      27.3  
8      13.9      31.1  
9      14.5      28.6  
10     13.8      28.1  
11     13.0      26.9  
12     16.1      26.7  
13     12.3      27.3  
14     15.0      29.7  
15     14.9      26.8  
16     13.8      28.3  
17     16.9      27.9  
18     15.0      26.4  
19     16.3      30.7  
20     13.3      28.3  
21     12.3      28.9  
22     14.8      31.5  
23     12.3      27.6  
24     14.6      27.0  
25     13.6      29.1  
26     14.0      28.3  
27     13.2      27.1  
28     13.5      29.1  
29     13.4      29.8

In [91]:
df['fgdiff'] = df['eFG%'] - df['Opp eFG%']
df['tovdiff'] = df['TOV%'] - df['OppTOV%']
df['ftdiff'] = df['FTARate'] - df['OppFTA']
df['rebdiff'] = df['OREB%'] - df['OppOREB%']

x = df[['fgdiff', 'tovdiff', 'ftdiff', 'rebdiff']]
y = df['win']

In [92]:
df

Team  win  eFG%  FTARate  TOV%  OREB%  Opp eFG%  OppFTA  \
0          Milwaukee Bucks   58  55.5    0.248  14.3   28.4      52.0   0.225   
1           Boston Celtics   57  56.6    0.243  13.3   25.5      52.8   0.234   
2       Philadelphia 76ers   54  56.3    0.300  13.9   25.8      54.1   0.277   
3           Denver Nuggets   53  57.3    0.259  14.7   28.9      54.3   0.260   
4      Cleveland Cavaliers   51  55.6    0.264  13.7   27.3      53.5   0.268   
5        Memphis Grizzlies   51  54.0    0.259  13.3   30.2      52.6   0.262   
6         Sacramento Kings   48  57.2    0.284  13.3   26.5      56.3   0.258   
7          New York Knicks   47  54.1    0.285  13.1   31.8      53.6   0.269   
8            Brooklyn Nets   45  56.2    0.260  13.9   23.4      53.0   0.276   
9             Phoenix Suns   45  53.5    0.241  13.6   30.7      53.2   0.298   
10   Golden State Warriors   44  57.1    0.224  15.8   27.7      54.0   0.278   
11             LA Clippers   44  55.1    0.278  14.3   27.3      54.3   0.253   
12              Miami Heat   44  53.0    0.270  13.8   27.4      56.1   0.253   
13      Los Angeles Lakers   43  54.2    0.299  13.7   26.9      53.5   0.222   
14  Minnesota Timberwolves   42  56.0    0.271  15.0   25.8      54.0   0.293   
15    New Orleans Pelicans   42  54.3    0.279  14.5   29.0      54.3   0.272   
16           Atlanta Hawks   41  54.1    0.244  12.6   28.9      55.2   0.257   
17         Toronto Raptors   41  51.7    0.257  11.8   31.6      56.5   0.280   
18           Chicago Bulls   40  55.0    0.251  13.3   23.6      54.4   0.253   
19   Oklahoma City Thunder   40  53.1    0.256  12.7   28.9      54.7   0.288   
20        Dallas Mavericks   38  56.5    0.298  12.4   22.6      54.9   0.290   
21               Utah Jazz   37  54.7    0.265  15.1   30.8      54.1   0.265   
22          Indiana Pacers   35  54.5    0.265  14.6   28.5      55.4   0.287   
23      Washington Wizards   35  55.0    0.258  14.2   27.3      54.0   0.244   
24           Orlando Magic   34  53.2    0.290  15.1   28.1      55.0   0.265   
25  Portland Trail Blazers   33  54.9    0.289  14.5   26.4      56.3   0.276   
26       Charlotte Hornets   27  51.6    0.261  13.9   27.6      54.4   0.266   
27         Houston Rockets   22  51.6    0.285  16.2   34.4      56.4   0.275   
28       San Antonio Spurs   22  52.5    0.229  14.8   29.9      57.6   0.260   
29         Detroit Pistons   17  52.0    0.295  15.0   29.2      55.7   0.298   

    OppTOV%  OppOREB%  fgdiff  tovdiff  ftdiff  rebdiff  
0      11.6      25.5     3.5      2.7   0.023      2.9  
1      12.6      25.4     3.8      0.7   0.009      0.1  
2      14.5      27.9     2.2     -0.6   0.023     -2.1  
3      13.6      27.3     3.0      1.1  -0.001      1.6  
4      16.1      28.5     2.1     -2.4  -0.004     -1.2  
5      14.8      28.9     1.4     -1.5  -0.003      1.3  
6      13.9      26.9     0.9     -0.6   0.026     -0.4  
7      12.9      27.3     0.5      0.2   0.016      4.5  
8      13.9      31.1     3.2      0.0  -0.016     -7.7  
9      14.5      28.6     0.3     -0.9  -0.057      2.1  
10     13.8      28.1     3.1      2.0  -0.054     -0.4  
11     13.0      26.9     0.8      1.3   0.025      0.4  
12     16.1      26.7    -3.1     -2.3   0.017      0.7  
13     12.3      27.3     0.7      1.4   0.077     -0.4  
14     15.0      29.7     2.0      0.0  -0.022     -3.9  
15     14.9      26.8     0.0     -0.4   0.007      2.2  
16     13.8      28.3    -1.1     -1.2  -0.013      0.6  
17     16.9      27.9    -4.8     -5.1  -0.023      3.7  
18     15.0      26.4     0.6     -1.7  -0.002     -2.8  
19     16.3      30.7    -1.6     -3.6  -0.032     -1.8  
20     13.3      28.3     1.6     -0.9   0.008     -5.7  
21     12.3      28.9     0.6      2.8   0.000      1.9  
22     14.8      31.5    -0.9     -0.2  -0.022     -3.0  
23     12.3      27.6     1.0      1.9   0.014     -0.3  
24     14.6      27.0    -1.8      0.5   0.025      1.1  
25

In [97]:
x_test = df[['fgdiff', 'tovdiff', 'ftdiff', 'rebdiff']]
y_test = df['win']

model.score(x_test, y_test)

0.7453873344275213

In [98]:
y_pred = model.predict(x_test)

print("Mean Squared Error:", mean_squared_error(y_test, y_pred))
print("R-squared:", r2_score(y_test, y_pred))

Mean Squared Error: 24.731376915940107
R-squared: 0.7453873344275213


##### Our regression model has an MSE of 24.73 and an R^2 value of 0.75. That means we can be ~95% sure that our predicted value for wins will be within 4.973 * 2 = 9.95 wins of the actual value